### introspection

In [1]:
b = [1, 2, 3]
help(b) # in vs code
# in trminal
# ipython
# In[1] : [1,2,3]
# In[2] :b?

Help on list object:

class list(object)
 |  list(iterable=(), /)
 |  
 |  Built-in mutable sequence.
 |  
 |  If no argument is given, the constructor creates a new empty list.
 |  The argument must be an iterable if specified.
 |  
 |  Methods defined here:
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __delitem__(self, key, /)
 |      Delete self[key].
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(...)
 |      x.__getitem__(y) <==> x[y]
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __iadd__(self, value, /)
 |      Implement self+=value.
 |  
 |  __imul__(self, value, /)
 |      Implement self*=value.
 |  
 |  __init__(self, /, *args, **kwargs)
 |      Initialize self.  See help(type(self)) for accurate sign

### function

In [2]:
# function call
def add(a, b):
    return a + b  # Returns the sum of a and b

result = add(3, 5)  # Calling the function with arguments 3 and 5
print(result)  # Output: 8

# object method call
text = "hello" # string object
print(text.upper()) # upper() is string method 

# variable stores refrences , not copies 

a = [1, 2, 3]  # 'a' points to a list object
b = a          # 'b' now also points to the same list
print(b)       # Output: [1, 2, 3]

# Changes in One Variable Affect the Other
a.append(4)
print(b)

# copy instead of refrencing 
a = [1, 2, 3]
# a.append(4) 
b = a.copy()   # Creates a new list with the same contents
a.append(4)    # Modifying 'a' will not affect 'b'

print(a)  # Output: [1, 2, 3, 4]
print(b)  # Output: [1, 2, 3]

# modifying a list inside a function (mutuable object)
def append_element(some_list, element):
    some_list.append(element)  # Modifies the original list

data = [1, 2, 3]
append_element(data, 4)

print(data)  # Output: [1, 2, 3, 4]
# what happen if the object is immutable 
def change_number(x):
    x = x + 1  # This creates a new variable 'x' inside the function

num = 10
change_number(num)

print(num)  # Output: 10 (original value remains unchanged)


8
HELLO
[1, 2, 3]
[1, 2, 3, 4]
[1, 2, 3, 4]
[1, 2, 3]
[1, 2, 3, 4]
10


### immutable and mutuable objects

In [1]:
# Mutable Objects (Can Be Changed)
'''Mutable List'''
a = [1, 2, 3]  # List is mutable
b = a  # 'b' references the same list

a.append(4)  # Modify the list
print(a)  # Output: [1, 2, 3, 4]
print(b)  # Output: [1, 2, 3, 4] (changes are reflected in 'b' too!)

'''mutable dictionary'''
person = {"name": "Alice", "age": 25}
person["age"] = 26  # Modifying dictionary

print(person)  # Output: {'name': 'Alice', 'age': 26}

# Immutable Objects (Cannot Be Changed)

'''Immutable Integer'''
x = 10
y = x  # 'y' gets the value of 'x'

x = x + 1  # This creates a NEW object for 'x'
print(x)  # Output: 11
print(y)  # Output: 10 (unchanged)
'''immutable string'''
s = "hello"
s = s.upper()  # Creates a new string

print(s)  # Output: "HELLO"

#Mutable vs Immutable: Function Behavior


'''Mutable Object in Function'''
def modify_list(lst):
    lst.append(4)  # Modifies the original list

numbers = [1, 2, 3]
modify_list(numbers)

print(numbers)  # Output: [1, 2, 3, 4]

'''Immutable Object in Function'''
def modify_number(n):
    n = n + 1  # Creates a new number object

num = 10
modify_number(num)

print(num)  # Output: 10 (unchanged)


[1, 2, 3, 4]
[1, 2, 3, 4]
{'name': 'Alice', 'age': 26}
11
10
HELLO
[1, 2, 3, 4]
10


### attributes and methods 

In [4]:
class Dog:
    def __init__(self, name):
        self.name = name  # 'name' is an attribute

my_dog = Dog("Bruno")
print(my_dog.name)  # Output: Bruno

class Dog:
    def __init__(self, name):
        self.name = name

    def bark(self):   # This is a method
        print(f"{self.name} says woof!")

my_dog = Dog("Bruno")
my_dog.bark()  # Output: Bruno says woof!


Bruno
Bruno says woof!


### duck type:
- Duck typing is a concept where the type or class of an object is less important than the methods or behaviors it has.
- if something acts like an iterable ,we treat it like one no need to check this type

#### example without duck type

In [ ]:
int add(int a, int b) {
    return a + b;
} # java

SyntaxError: invalid syntax (2250713675.py, line 1)

#### example with duck typing

In [6]:
def add(a, b):
    return a + b
add(2, 3)        # 5
add("Hi ", "there")  # "Hi there"
add([1, 2], [3])     # [1, 2, 3]


[1, 2, 3]

#### iterable
- anything you can use in a for loop is iterable 

In [7]:
# All of these are iterables
my_list = [1, 2, 3]
my_string = "hello"
my_tuple = (4, 5, 6)
my_dict = {'a': 1, 'b': 2}
my_set = {10, 20, 30}
for x in my_list:
    print(x)

for ch in my_string:
    print(ch)


1
2
3
h
e
l
l
o


### import modules and packages 
##### module :
- A module in Python is just a .py file that contains functions, variables, and classes that you can reuse in other Python files.
- example 

In [8]:
# some_module.py

PI = 3.14159

def f(x):
    return x + 2

def g(a, b):
    return a + b

import and intire module

In [11]:
import some_module

result = some_module.f(5)       # Calls f from some_module => 5 + 2 = 7
print(result)
pi = some_module.PI             # Access PI => 3.14159
print(pi)


7
3.14159


 Import Specific Items

In [13]:
from some_module import g, PI

result = g(5, PI)     # Calls g => 5 + 3.14159 = 8.14159
print(result)


8.14159


Using as to Rename Imports

In [14]:
import some_module as sm
from some_module import PI as pi, g as gf

r1 = sm.f(pi)         # sm.f is f from some_module => pi + 2 = 5.14159
r2 = gf(6, pi)        # gf is g from some_module => 6 + pi = 9.14159
print(r1,r2)


5.14159 9.14159


### binary operators and comparisons

##### binary operators :
- A binary operator is an operator that operates on two operands (left and right) to perform an operation.
- example = + , - , * , / , // , **

##### comparison operator :
- A comparison operator is used to compare two values.
It returns a Boolean value: either True or False.
example (== , != , > , < , >= , <=)

##### logical / bitwise operator:
- for booleans(true/false):
- and / & = return true when both are true 
- or = return false at least one is true 
- ^ / XOR = return true only one is true 

is VS (==):
- == : value equality(same content)
- is == : identity(same object in memory)

In [15]:
a = [1, 2, 3]
b = a         # b points to the *same* list as a
c = list(a)   # c is a *copy* of a

print(a == c)     # True  (same content)
print(a is c)     # False (different memory)

print(a is b)     # True  (both point to same object)


True
False
True


is none / is not

In [16]:
a = None

if a is None:
    print("a is empty or unassigned")


a is empty or unassigned


### scalar type :
- Scalar types are the basic building blocks of data in Python — they represent a single value at a time.

### here are main scalar type you will see in python:

- #### Type	     Description
- None  &nbsp;	     Represents “nothing” or “no value”.
  
- str	  &nbsp;      String type — holds text (like "hello").
  
- bytes	  &nbsp;    Holds raw binary data (not commonly used in beginner data analysis).
  
- float	  &nbsp;   Floating point number (e.g. 3.14, 0.0001, -99.99).
  
- bool	  &nbsp;   Boolean value: True or False.
  
- int	  &nbsp;       Integer (whole number), and it can be as big as needed (no size limit).
  
  

### bytes and unicode:

- # 1. unicode ?
  - Unicode is a standard for representing characters from almost all languages.

  - In Python 3, all strings (str) are Unicode by default.

In [1]:
# example 
val = "español"
print(val)  # Output: español


español


- # 2. bytes ?
  
  - bytes are raw binary data - not text 
  
  - You use bytes when working with things like files, images, or network data.  

In [ ]:
# converting unicode to bytes 
val_utf8 = val.encode("utf-8")
print(val_utf8)  

# and also you can go back bytes to unicode
val_utf8.decode("utf-8")  



b'espa\xc3\xb1ol'


'español'

### Encoding and Decoding 

- # 1 encoding :
  
   - converting strings into bytes 
   - <pre>      text = "hello"
           encoded = text.encode("utf-8")  # 🔄 Encoding text to bytes
           print(encoded)  # Output: b'hello' </pre>

- # 2 decoding :
  
   - converting bytes again into string
   - <pre>      data = b"hello"
           decoded = data.decode("utf-8")  # 🔄 Decoding bytes to text
           print(decoded)  # Output: hello  </pre>



### Booleans :

- A Boolean is a type of data that can hold only two values:

- Why Are Booleans Important?
   - We use booleans in:

       - if statements

       - while loops

       - filters (df[df['age'] > 30])

       - comparisons (==, !=, <, etc.)
  
- Booleans are super useful when filtering data:



In [1]:

import pandas as pd


df = pd.DataFrame({"age": [15, 22, 34, 17]})
df[df["age"] >= 18]  # Filters only adults (age >= 18)


age
1   22
2   34

### date and time module :

- The datetime module provides classes for manipulating dates and times in a simple and flexible way.

In [ ]:
from datetime import datetime, date, time
# creating a date and time object :
dt = datetime(2023, 5, 11, 14, 30, 0)
print(dt)
# Year: 2023, Month: May (5), Day: 11, Hour: 14 (2 PM), Minute: 30, Second: 0
# for accessing component
dt.year    # 2023
dt.month   # 5
dt.day     # 11
dt.minute  # 30




2023-05-11 14:30:00


30

#### operations with date and time :

- You can do arithmetic with datetime objects using timedelta:

In [5]:
from datetime import timedelta

dt + timedelta(days=5)
# Adds 5 days

dt - timedelta(hours=2)
# Subtracts 2 hours


datetime.datetime(2023, 5, 11, 12, 30)

##### timedelta :

- which is used to perform arithmetic with dates and times.

In [ ]:
from datetime import datetime, timedelta

dt1 = datetime(2011, 11, 15, 22, 30)
dt2 = datetime(2011, 10, 29, 20, 30, 21)

delta = dt1 - dt2
print(delta)


new_time = dt2 + delta
print(new_time)




17 days, 1:59:39
2011-11-15 22:30:00


### built in sequence funtion in pyton:

- sequences are ordered collections of elements, and Python provides several built-in functions that can be used to work with these sequences (like lists, tuples, strings, etc.).

### Here are some important built-in sequence functions in Python:

# 1. min() 

- return the smallest number in the list 

In [8]:
numbers = [10, 20, 5, 8]
print(min(numbers))  # Output: 5


5


# 2. max() 

- return the largest number in the list 

In [9]:
numbers = [10, 20, 5, 8]
print(max(numbers))  # Output: 20


20


# 3. sum() 

- return all the sum of the all item in the sequence 

In [10]:
numbers = [1, 2, 3, 4]
print(sum(numbers))  # Output: 10


10


# 4. sorted() :

- Returns a new sorted list from the elements of any sequence (like a list, tuple, string)


In [11]:
numbers = [3, 1, 4, 2]
print(sorted(numbers))  # Output: [1, 2, 3, 4]

[1, 2, 3, 4]


# 5. enumerate():

-  Adds a counter to an iterable and returns it as an enumerate object
  
- This can be very helpful when you need both the index (position) and the value of each element in a sequence while iterating through it.

In [ ]:
fruits = ['apple', 'banana', 'cherry']
for index, fruit in enumerate(fruits):
    print(index, fruit)


0 apple
1 banana
2 cherry


# 6 . zip() :

- Combines two or more sequences (lists, tuples, etc.) into a single iterable of tuples.

In [13]:
names = ['Alice', 'Bob', 'Charlie']
scores = [85, 90, 78]
result = zip(names, scores)
print(list(result)) 


[('Alice', 85), ('Bob', 90), ('Charlie', 78)]


# 7. all():

-  Returns True if all elements of the sequence are truthy (i.e., not False, 0, None, etc.), otherwise returns False.

In [16]:
numbers = [1, 2, 3]
print(all(numbers))  
char = [1,2,3,"hello",0]
print(all(char))

True
False


# 8 . any():

-  Returns True if any element of the sequence is truthy, otherwise returns False.

In [ ]:
numbers = [0, 2, 3]
print(any(numbers))  
numbers = [0, False, None, "", []]

result = any(numbers)
print(result)  



True
False


### list comprehension :

- List comprehension is a shorter and cleaner way of creating a new list by performing an operation on each item in an existing list (or any iterable, like a tuple or string). It’s like a one-liner for creating lists instead of using loops.

- why it is useful ?

   - Normally, to create a list with some modifications or filters, you might use a for loop. List comprehension helps you do that in one line without having to write more code.

- example 1 :

In [ ]:
numbers = [1, 2, 3, 4, 5]
squares = [x**2 for x in numbers]
print(squares)


- example 2(using condition):

 

In [18]:
numbers = [1, 2, 3, 4, 5, 6]
even_numbers = [x for x in numbers if x % 2 == 0]
print(even_numbers)


[2, 4, 6]


- modifying element based on condition 

In [1]:
numbers = [1, 2, 3, 4, 5, 6]
modified_numbers = [x*2 if x % 2 == 0 else x*3 for x in numbers]
print(modified_numbers)


[3, 4, 9, 8, 15, 12]


### set comprehension :

- Set comprehension is a concise and readable way to create a set from an existing iterable (like a list, tuple, or string). It’s very similar to list comprehension, but it creates a set instead of a list. A set is an unordered collection of unique elements, which means that duplicates are automatically removed.

- example 1 :

In [2]:
numbers = [1, 2, 3, 4, 5]
squares_set = {x**2 for x in numbers}
print(squares_set)


{1, 4, 9, 16, 25}


- example 2 (using condition) :

In [ ]:
numbers = [1, 2, 3, 4, 5, 6]
even_numbers_set = {x for x in numbers if x % 2 == 0}
print(even_numbers_set)


- example 3 :

Modifying Elements in Set Comprehension
Let’s say we want to create a set where all the odd numbers are doubled and the even numbers are kept as they are.

In [3]:
numbers = [1, 2, 3, 4, 5, 6]
modified_set = {x*2 if x % 2 != 0 else x for x in numbers}
print(modified_set)


{2, 10, 4, 6}


- example 4 :

removing duplicates automatically

In [4]:
numbers = [1, 2, 3, 4, 4, 5, 5, 6]
unique_set = {x for x in numbers}
print(unique_set)


{1, 2, 3, 4, 5, 6}


### dict comprehension :

- Dictionary comprehension is a concise way to create dictionaries in Python. It's similar to list comprehension and set comprehension, but instead of creating a list or a set, you create a dictionary with key-value pairs. It allows you to easily apply transformations, filters, or conditions to the data as you build the dictionary.